### PR TITLE
Audit Logs and Reasons

### PR DESCRIPTION
**Pull request by MinnDevelopment**
This added:
-  `AuditableRestAction` / `AuditableRestAction.EmptyRestAction`
    Allowing to add a reason to any auditable request via `reason(String)`
- New reason routes for kick/ban to be able to use in case client accounts may not be able to use the `X-Audit-Log-Reason` header
- `AuditLogEntry` / `AuditLogPaginationAction`
    Allowing to view and iterate over all audit log entries via pagination
- `AuditLogChange`
    POJO containing changes for a single field/key 
- `ActionType` / `TargetType`
   Identifies what kind of action an `AuditLogEntry` represents

This directly addresses #344

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/core/entities/impl/TextChannelImpl.java [hunk 1]
```diff
@@ -132,15 +94,54 @@ protected void handleResponse(Response response, Request<List<Webhook>> request)
...
+                if (response.isOk())
+                    request.onSuccess(null);
+                else
+                    request.onFailure(response);
+            }
+        };
+    }
+
+    @Override
+    public AuditableRestAction<Void> deleteWebhookById(String id)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/entities/impl/TextChannelImpl.java]: Why did all of this get shifted? O.o

File: src/main/java/net/dv8tion/jda/core/managers/GuildController.java [hunk 2]
```diff
@@ -377,19 +387,25 @@ protected void handleResponse(Response response, Request<Integer> request)
      *             <br>See {@link net.dv8tion.jda.core.utils.PermissionUtil#canInteract(Member, Member) PermissionUtil.canInteract(Member, Member)}</li>
      *         </ul>
      *
-     * @return {@link net.dv8tion.jda.core.requests.RestAction RestAction}
+     * @return {@link net.dv8tion.jda.core.requests.restaction.AuditableRestAction AuditableRestAction}
      *         Kicks the provided Member from the current Guild
      */
-    public RestAction<Void> kick(Member member)
+    public AuditableRestAction<Void> kick(Member member, String reason)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/managers/GuildController.java]: Is this reason different from the AuditLog reason?

Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/managers/GuildController.java]: Client accounts cannot use the `X-Audit-Log-Reason` header that is why I added kick/ban with the query parameter.

**Comment by MinnDevelopment** [comment]
I believe this is now ready to get merged.

**Pull request merged**